teps/0091-trusted-resources.md

```diff
@@ -313,12 +418,119 @@ What other approaches did you consider and why did you rule them out?  These do
+
+Option 4 is less preferred than 3 because the mutating webhook would need to contain verification logic, and this is not well aligned with the intention of mutating webhooks.
+
+### Options for storing the signature
+
+In this TEP we propose to store the signature in the `annotation` of the resource because of the easy implementation. Alternatives also include storing the signature separately alongside with the resource or in other remote sources. This could be future work after this TEP.
+
+### Options for storing the public keys
+
+Option 1: Configure the keys in a dedicated CRD. This can help to support multiple keys in a central place. This is the proposed option in this TEP.
+
+Option 2: Add a new field `keyref` to refer to the key path
+```yaml
+kind: TaskRun
+apiVersion: tekton.dev/v1beta1
+metadata:
+ name: example-tr
+spec:
+ taskRef:
+   name: example-task
+   keyref: keypath # this defines the public key of this taskref
+```
+
+```yaml
+kind: PipelineRun
+apiVersion: tekton.dev/v1beta1
+metadata:
+ name: example-pr
+spec:
+ pipelineRef:
+   name: example-pipeline
+   keyref: keypath # this defines the public key of this pipelineref
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+ annotations:
+   tekton.dev/signature: MEQCIHhFC480mv7mh/6dxAlp/mvuvGXanuSghMsT+iBhWxt5AiBDvHv8sfKjJ3Ozrzvp+cjr30AOx1SPQDCcaJjlpAoVrA==
+ name: example-Pipeline
+spec:
+ tasks:
+   - name: example-task
+     taskRef:
+       name: example-task1
+       keyref: keypath1  # this defines the public key of this taskref
+   - name: example-task2
+     taskRef:
+       name: example-task2
+       keyref: keypath2  # this defines the public key of this taskref
+```
+`keyref` defines the path of the public key of this taskRef. All the tasks within a pipeline need to be signed off, and they may come from different sources with different public keys.
+So a `keyref` is needed when signing the pipeline. The signing cli doesn't need to verify the referred keys when signing the pipeline.
+
+**Pros vs Cons:**
+| Option | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Config at install level (e.g. at CRD) | Easy to config at a central place | Need to loop all the keys for each verification
+| Add keyref in taskRef&pipelineRef | Each resource can specify the corresponding key, no need to loop all available keys for one resource's verification| Introduce cost for users to write and maintain YAML files
+
+### Options for integrating with remote resolution
+
+| Method | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Fetch remote resources in validating Webhook | Verification fail fast in webhook | Duplicate work for Resolution, may introduce latency, and one extreme case is that the resource verified may not the the same in reconciler's resolved resource
```

Comment:
Fine to leave out of this proposal for now (doesn't modify the rest of it), but I think there may be some value in still having validating webhook checks at admission for API resources without resolution (e.g. just validate that the message we received was unmodified) on top of the resolve-at-runtime behavior.

It would be useful to be able to reject unsigned resources at creation to get quick feedback on whether the signature is correct. We'd still check this again at runtime to verify the policy hasn't changed, but it can help circuit bad configs faster.

Reply:
Yeah that makes sense!